DOCSP-46328: Django connection configuration (#142)
mongodb/docs-pymongo: source/configure-connection.txt

.. _django-connection-configuration:

==================================
Configure Your Database Connection
==================================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, settings

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to connect your Django project to MongoDB
and configure the connection.

Connection Configuration 
------------------------

After installing {+django-odm+} and creating a project, you can configure
your connection to MongoDB in the following ways:

- :ref:`django-connection-configure-manual` by updating your
  project's ``DATABASES`` setting
- :ref:`django-connection-configure-automatic` by using
  the ``parse_uri()`` method

.. tip::

   To learn how to install the Django ODM and create a

>    Django project, visit the :ref:`django-get-started` tutorial.

Author: Note: this link is broken until #132 is merged

.. _django-connection-configure-manual:

Manually Configure Database Settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To manually configure your connection to MongoDB, update
the ``DATABASES`` setting in your project's ``settings.py``
file. The ``DATABASES`` setting has a ``default`` key,
which you must configure to set MongoDB as the default database
connection.

To configure the ``default`` key, include the following

> fields:

Author: Note to tech reviewers: not sure if "field" is the right word here? are these also keys?

Reviewer: Yes `DATABASES` is a nested dictionary of settings so you might say

> To configure the ``default`` key, include the following nested `keys`.

Or

> To configure the ``default`` key, include the following settings.

Or

> To configure the ``default`` database, include the following settings.

But you probably wouldn't say "fields".

.. _django-connection-configuration:

==================================
Configure Your Database Connection
==================================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, settings

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to connect your Django project to MongoDB
and configure the connection.

Connection Configuration 
------------------------

After installing {+django-odm+} and creating a project, you can configure
your connection to MongoDB in the following ways:

> - :ref:`django-connection-configure-manual` by specifying the

Author: I think they are full (imperative) sentences so I'll add periods

  ``DATABASES`` variable in your project's settings
- :ref:`django-connection-configure-automatic` by using
  the ``parse_uri()`` method

.. tip::

   To learn how to install the Django ODM and create a
   Django project, visit the :ref:`django-get-started` tutorial.

.. _django-connection-configure-manual:

Manually Configure Database Settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To manually configure your connection to MongoDB, update
the ``DATABASES`` variable in your project's ``settings.py``
file. Set the ``DATABASES`` variable to a dictionary value containing
the database settings. You must configure the dictionary's
``default`` key to set MongoDB as the default database
connection.

To configure the ``default`` key, assign a nested dictionary as its value.

> This nested dictionary has the following keys:

Author: That makes sense, reworked this part!

.. list-table::
   :header-rows: 1
   :widths: 20 80

   * - Key
     - Description

   * - **ENGINE**
     - The backend driver to use for the connection. Set this key to ``"django_mongodb_backend"``.

   * - **HOST**
     - | Your connection URI. For localhost connections, this key is optional. 
       | For SRV connections, you must include a scheme prefix (``mongodb+srv://``).
       |

>        | If connecting to a replica set or sharded cluster with multiple hosts, specify 
>        | each host separated by a comma.

Reviewer 2: s: slight wording suggestion. could also be helpful to include an example, either here or later.

```suggestion
       | To specify more than one host, include all hostnames in one string. Use
       | a comma to separate each hostname.
```

   * - **NAME**
     - The database you want to use.

   * - **USER**

>      - The username for authenticating to the database.

Reviewer 3: You can add if your connection requires no auth you do not need to provide it

   * - **PASSWORD**
     - The password for your database user.

>    * - **PORT**

Reviewer 3: Here you can say the default port is 27017

     - | The port number on which the database server is listening. 
       | For MongoDB Atlas connections, this key is optional.

   * - **OPTIONS**
     - A dictionary of additional connection options for the database. This key is optional.

.. _django-manual-config-example:

Example 
```````

> This example specifies the ``DATABASES`` variable to connect
> to a MongoDB deployment with the following configuration:

Reviewer 2: s:

```suggestion
In this example, the ``DATABASES`` variable specifies the
following connection configuration:
```

> - Connects to the ``my_database`` database

Reviewer 2: q: related to the preceding paragraph: who is the actor in the list? the client, ``DATABASES``, the
connection, the deployment, the app, or something else? It seems to mix client/app behavior ('Connects to the database' with config behavior ('Sets the connection option...')

Author: I was thinking DATABASES. I'll change some of the wording

Reviewer: Wow, English discussion going on here. I would say `s/variable to connect to/variable used to connect to/` and reserve further English language usage comments. 😄

- Provides authentication information for a database user
  whose username is ``my_user`` and password is ``my_password``
- Uses the default MongoDB port (``27017``)
- Sets the ``retryWrites`` connection option to ``true``,
  which configures the driver to automatically retry certain
  write operations if they fail
- Sets the ``w`` connection option to ``majority``,
  which configures the driver to wait for acknowledgement from a majority
  of replica set members before performing write operations

.. code-block:: python

    DATABASES = {
        "default": {
            "ENGINE": "django_mongodb_backend",
            "HOST": "mongodb+srv://cluster0.example.mongodb.net",
            "NAME": "my_database",
            "USER": "my_user",
            "PASSWORD": "my_password",
            "PORT": 27017,
            "OPTIONS": {
                "retryWrites": "true",
                "w": "majority",
            },
        },
    }

> .. tip::

Reviewer 2: s: move this to the OPTIONS row above (and take out of tip box to avoid nested admonition)

    To see a full list of connection options that you
    can set in the ``OPTIONS`` key, see the optional
    parameters for `MongoClient <https://pymongo.readthedocs.io/en/4.10.1/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient>`__
    in the PyMongo API documentation.

.. _django-connection-configure-automatic:

Automatically Configure Database Settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To automatically construct the ``DATABASES`` setting that configures

Reviewer 2: q: does this literally construct the DATABASES variable in that file with that same structure? or does it just have the same effect on the client/connection?

Author: The Django MongoDB README says it "constructs a DATABASES setting equivalent to the first example" -- @Jibola is it accurate to say that parse_uri() automatically constructs this setting?

Reviewer: `parse_uri` is a convenience method so if it helps to leave it out for clarity's sake then I would do that. That said, `DATABASES` has to exist and it has to have the required keys. For example, I can run Django without that setting but if I try to access the admin I get
```
settings.DATABASES is improperly configured. Please supply the ENGINE value. 
Check settings documentation for more details.
```

Also FWIW my preference for testing is currently: https://github.com/aclark4life/project-templates/blob/main/project_template/project_name/settings/__init__.py#L7 with `MONGODB_URI` set by direnv.

Author: It seems accurate to say it's constructing the DATABASES setting, so I'll leave that in. 
Since I'm showing how to configure the same settings as the previous example, I'll leave the MONGODB_URI variable that shows how to do that

your MongoDB connection, you can use the ``parse_uri()`` method. This
method accepts the following arguments:

- ``uri``: Your MongoDB connection URI.
- ``conn_max_age``: Configures persistent database connections.
  This argument is optional. To learn more, see 
  `Persistent connections <https://docs.djangoproject.com/en/stable/ref/databases/#persistent-database-connections>`__
  in the Django documentation.
- ``test``: Provides a dictionary of settings for test
  databases. This argument is optional. To learn more, see 
  `the TEST setting <https://docs.djangoproject.com/en/stable/ref/settings/#test>`__
  in the Django documentation.

Example 
```````

> The following example uses the ``parse_uri()`` method to connect
> to a MongoDB deployment with the same configuration as 

Reviewer 2: s:

```suggestion
The following example uses the ``parse_uri()`` method to specify
the same connection configuration as 
```

Reviewer: NIT: s/method/function/

the previous :ref:`manual configuration <django-manual-config-example>`
example:

.. code-block:: python

    import django_mongodb_backend

    MONGODB_URI = "mongodb+srv://my_user:my_password@cluster0.example.mongodb.net/my_database?retryWrites=true&w=majority"
    DATABASES["default"] = django_mongodb_backend.parse_uri(MONGODB_URI)

Additional Information
----------------------

To view a sample project that configures a MongoDB database connection,
see the :ref:`django-get-started-connect` step in the Getting Started
tutorial.

> To learn more about Django settings, see `Settings <https://docs.djangoproject.com/en/stable/ref/settings/>`__

Reviewer 3: NIT for future version mappings. We will need to have the links point to the correlating version. Right now it points to stable

.. _django-connection-configuration:

==================================
Configure Your Database Connection
==================================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, settings

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to connect your Django project to MongoDB

> and configure the connection.

Reviewer: FWIW this still reads "funny" to me … if you are connecting your Django project to MongoDB then you've already configured the connection. Feel free to override me because I can be NIT-happy.

Author: Reworded

Connection Configuration 
------------------------

After installing {+django-odm+} and creating a project, you can configure
your connection to MongoDB in the following ways:

- :ref:`django-connection-configure-manual` by specifying the
  ``DATABASES`` variable in your project's settings.
- :ref:`django-connection-configure-automatic` by using
  the ``parse_uri()`` method.

.. tip::

>    To learn how to install the Django ODM and create a

Reviewer: Should this be {+django-odm+} instead of "Django ODM" ?

   * - **USER**
     - The username for authenticating to the database, if your connection
       requires authentication.

   * - **PASSWORD**
     - The password for your database user, if your connection requires authentication.

   * - **PORT**
     - | The port number on which the database server is listening. The default
         port is ``27017``.
       | For MongoDB Atlas connections, this key is optional.

   * - **OPTIONS**
     - | A dictionary of additional connection options for the database. This key is optional.
       | To see a full list of connection options that you can set in the ``OPTIONS`` key,
         see the optional parameters for `MongoClient <https://pymongo.readthedocs.io/en/4.10.1/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient>`__
         in the PyMongo API documentation.

.. _django-manual-config-example:

Example 
```````

In this example, the ``DATABASES`` variable specifies the
following connection configuration:

- Sets the database to ``my_database``
- Provides authentication information for a database user
  whose username is ``my_user`` and password is ``my_password``
- Specifies the default MongoDB port (``27017``)

> - Sets the ``retryWrites`` connection option to ``true``,

Reviewer 2: i: there's a double verb here--'specifies' in the paragraph and then 'connects'/'sets' in the list items. i can't add a suggestion in GH, but maybe something like this:

In this example, the ``DATABASES`` variable performs the
following steps:

Author: changed to just "...performs the following actions"

.. tip::

    To see a full list of connection options that you

>     can set in the ``OPTIONS`` nested key, see the optional

Reviewer: s/key/dictionary/, otherwise, nice!
